ambry-server/src/test/java/com.github.ambry.server/Http2BlockingChannel.java

```diff
@@ -0,0 +1,136 @@
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http2.DefaultHttp2DataFrame;
+import io.netty.handler.codec.http2.DefaultHttp2Headers;
+import io.netty.handler.codec.http2.DefaultHttp2HeadersFrame;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2StreamChannel;
+import io.netty.handler.codec.http2.Http2StreamChannelBootstrap;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A HTTP2 implementation of {@link ConnectedChannel}. This implementation is for test now. It will be imporved to
+ * to support replication in the future.
+ */
+public class Http2BlockingChannel implements ConnectedChannel {
+  private static final Logger logger = LoggerFactory.getLogger(Http2BlockingChannel.class);
+  private final Http2ResponseHandler http2ResponseHandler;
+  private final String hostName;
+  private final int port;
+  private EventLoopGroup workerGroup;
+  private Channel channel;
+
+  public Http2BlockingChannel(String hostName, int port) {
+    http2ResponseHandler = new Http2ResponseHandler();
+    this.hostName = hostName;
+    this.port = port;
+  }
+
+  @Override
+  public void connect() throws IOException {
+    workerGroup = new NioEventLoopGroup();
+    Bootstrap b = new Bootstrap();
+    b.group(workerGroup);
+    b.channel(NioSocketChannel.class);
+    b.option(ChannelOption.SO_KEEPALIVE, true);
+    b.remoteAddress(hostName, port);
+    SSLFactory sslFactory;
+    try {
+      sslFactory = new NettySslHttp2Factory(new SSLConfig(new VerifiableProperties(new Properties())));
+    } catch (GeneralSecurityException e) {
+      throw new IOException(e);
+    }
+    b.handler(new Http2ClientChannelInitializer(sslFactory, hostName, port));
+
+    // Start the client.
+    channel = b.connect().syncUninterruptibly().channel();
+    logger.info("Connected to remote host");
+  }
+
+  @Override
+  public void disconnect() throws IOException {
+    channel.disconnect().syncUninterruptibly();
+    workerGroup.shutdownGracefully();
+  }
+
+  @Override
+  public void send(Send request) throws IOException {
+    ByteBufferChannel byteBufferChannel = new ByteBufferChannel(ByteBuffer.allocate((int) request.sizeInBytes()));
+    while (!request.isSendComplete()) {
+      request.writeTo(byteBufferChannel);
+    }
+    byteBufferChannel.getBuffer().position(0);
+    ByteBuf byteBuf = Unpooled.wrappedBuffer(byteBufferChannel.getBuffer());
+
+    Http2ClientStreamInitializer initializer = new Http2ClientStreamInitializer(http2ResponseHandler);
+    Http2StreamChannel childChannel =
+        new Http2StreamChannelBootstrap(channel).handler(initializer).open().syncUninterruptibly().getNow();
```

Comment:
this is weird given that we have initialized the handler in the connect method. It seems like all we need from connect method is just a functional channel so we can build a http2 stream handler on top of that channel?

Reply:
yes, http2 stream channel is a child channel of http2 connection channel.